### PR TITLE
Event-based node components

### PR DESCRIPTION
**Pull request by art-gor**
### Description of the Change

Event-based node components

### Benefits

Loose coupling between components: testability, maintainability, bus factor, etc.

### Possible Drawbacks 

None

### CODE REVIEW

File: core/node/pubsub_gate.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+      return false;
+    }
+
+    std::error_code e;
+
+    auto cid_res = common::getCidOf(raw);
+    if (!cid_res) {
+      e = cid_res.error();
+    } else {
+      bool decoding_secp_msg = (raw.data()[0] == kCborTwoElementsArrayHeader);
```
Comment (turuslan) [inline core/node/pubsub_gate.cpp:126]: Pubsub message is `SignedMessage` (both bls and secp)

File: core/node/blocksync_common.cpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/**
```
Comment (turuslan) [inline core/node/blocksync_common.cpp:1]: This file does nothing and should be removed

File: core/node/head_constructor.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    auto &creator = candidates_[TipsetKey::hash(header.parents)];
+    auto res = creator.canExpandTipset(header);
+    if (!res) {
+      log()->debug("cannot expand tipset with new block, {}",
+                   res.error().message());
+    }
+    res = creator.expandTipset(block_cid, header);
+    if (!res) {
+      if (!res) {
```
Comment (turuslan) [inline core/node/head_constructor.cpp:95]: `expandTipset` will `assert`, fix `if`

File: core/node/events.hpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    PeerId from;
+    CID block_cid;
+    BlockWithCids block;
+  };
+
+  struct MessageFromPubSub {
+    PeerId from;
+    CID cid;
+    UnsignedMessage msg;
+    boost::optional<Signature> signature;
```
Comment (turuslan) [inline core/node/events.hpp:53]:
```suggestion
    SignedMessage msg;
```

File: core/node/main/config.cpp [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    constexpr int kDefaultPort = 2000;
+    constexpr std::string_view kDefaultConfigFileName = "default.cfg";
+
+    const std::string &getLocalIP() {
+      static const std::string a = [] {
+        boost::asio::io_context io;
+        boost::asio::ip::tcp::resolver resolver(io);
+        boost::asio::ip::tcp::resolver::query query(
+            boost::asio::ip::host_name(), "");
+        boost::asio::ip::tcp::resolver::iterator it = resolver.resolve(query);
```
Comment (turuslan) [inline core/node/main/config.cpp:28]: Throws exception sometimes
```suggestion
        boost::system::error_code ec;
        boost::asio::ip::tcp::resolver::iterator it = resolver.resolve(query, ec);
```

File: core/node/main/builder.cpp [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+
+      const auto &roots = result.value();
+      if (roots.empty()) {
+        return Error::NO_GENESIS_BLOCK;
+      }
+
+      if (config.genesis_cid) {
+        if (config.genesis_cid.value() != roots[0]) {
+          log()->error("Genesis mismatch: got cids:{}, expected:{}",
+                       fmt::join(toStrings(roots), " "),
```
Comment (turuslan) [inline core/node/main/builder.cpp:113]:
```suggestion
                       fmt::join(roots, " "),
```

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: core/node/blocksync_request.cpp [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+          }
+          block_stored(std::move(block_cid), std::move(header));
+        }
+      }
+
+      return outcome::success();
+    }
+
+    // XXX
+    static int xxx = 0;
```
Comment (Alexey-N-Chernyshov) [inline core/node/blocksync_request.cpp:221]: What is it?

File: core/node/head_constructor.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    res = creator.expandTipset(block_cid, header);
+    if (!res) {
+      log()->error("cannot expand tipset with new block, {}",
+                     res.error().message());
+      return;
+    }
+
+    log()->debug("new possible head");
+
+    // XXX
```
Comment (Alexey-N-Chernyshov) [inline core/node/head_constructor.cpp:107]: Add todo

File: core/node/blocksync_request.cpp (hunk 7 above)
Comment (turuslan) [reply core/node/blocksync_request.cpp:221]: Logging reference count when objects are created and destoryed

**Review by turuslan — APPROVED** [review]

**Pull request merged**